lib/evmone/instructions.hpp

```diff
@@ -925,6 +925,59 @@ inline Result mcopy(StackTop stack, int64_t gas_left, ExecutionState& state) noe
     return {EVMC_SUCCESS, gas_left};
 }
 
+inline Result dataload(StackTop stack, int64_t gas_left, ExecutionState& state) noexcept
+{
+    auto& index = stack.top();
+
+    if (state.data.size() < 32 || (state.data.size() - 32) < index)
+        return {EVMC_INVALID_MEMORY_ACCESS, gas_left};  // TODO: Introduce EVMC_INVALID_DATA_ACCESS
+
+    const auto begin = static_cast<size_t>(index);
+    index = intx::be::unsafe::load<uint256>(&state.data[begin]);
+    return {EVMC_SUCCESS, gas_left};
+}
+
+inline void datasize(StackTop stack, ExecutionState& state) noexcept
+{
+    stack.push(state.data.size());
+}
+
+inline code_iterator dataloadn(StackTop stack, ExecutionState& state, code_iterator pos) noexcept
+{
+    const auto index = read_uint16_be(&pos[1]);
+
+    const auto begin = static_cast<size_t>(index * 32);
+    stack.push(intx::be::unsafe::load<uint256>(&state.data[begin]));
+    return pos + 3;
+}
+
+inline Result datacopy(StackTop stack, int64_t gas_left, ExecutionState& state) noexcept
+{
+    const auto& mem_index = stack.pop();
+    const auto& data_index = stack.pop();
+    const auto& size = stack.pop();
+
+    if (!check_memory(gas_left, state.memory, mem_index, size))
+        return {EVMC_OUT_OF_GAS, gas_left};
+
+    auto s = static_cast<size_t>(size);
+
+    if (data_index + s > state.data.size())
```

Comment:
`data_index + s` can overflow.

Reply:
True, I wrongly changed it after this comment https://github.com/ethereum/evmone/pull/586#discussion_r1192195251